File: translation_output/codellama/codenet/python/java/p00345.java
<s> [INST] <<SYS>>
You are an expert java programmer and assistant
<</SYS>>

python code:

def gcd(x, y):
    return gcd(y, x%y) if y else x

def printV(x, y):
    g = gcd(x, y)
    print(str(x//g) + "/" +  str(y//g))

S = input()
all = ""
sub = ""
p = -1

for i in range(len(S)):
    if S[i] == '.':
        o = i
    elif S[i] == '(':
        p = i
        sub = all
    elif S[i] != ')':
        all += S[i]

d = len(S) - o - 1
l = p - o - 1
if p == -1:
    printV(int(all), 10**d)
else:
    d -= 2 # for ()
    printV(int(all) - int(sub), 10**d - 10**l)



Translate the above python code to java.

java code:

[/INST]

Here is the Java code equivalent to the Python code:
```java
import java.util.Scanner;

public class GCD {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String input = scanner.nextLine();
        int o = input.indexOf('.');
        int p = input.indexOf('(');
        int d = input.length() - o - 1;
        int l = p - o - 1;
        String all = "";
        String sub = "";
        for (int i = 0; i < input.length(); i++) {
            if (input.charAt(i) == '.') {
                o = i;
            } else if (input.charAt(i) == '(') {
                p = i;
                sub = all;
            } else if (input.charAt(i) != ')') {
                all += input.charAt(i);
            }
        }
        if (p == -1) {
            printV(Integer.parseInt(all), 10**d);
        } else {
            d -= 2; // for ()
            printV(Integer.parseInt(all) - Integer.parseInt(sub), 10**d - 10**l);
        }
    }

    public static void printV(int x, int y) {
        int g = gcd(x, y);
        System.out.println(x / g + "/" + y / g);
    }

    public static int gcd(int x, int y) {
        return y == 0 ? x : gcd(y, x % y);
    }
}
```
Note that the `gcd` function is implemented using the recursive approach, and the `printV` function is used to print the result in the desired format. The `main` function reads the input from the user, parses it, and calls the `printV` function to print the result.</s>

File: translation_output/codellama/codenet/java/python/p00345.py
<s> [INST] <<SYS>>
You are an expert python programmer and assistant
<</SYS>>

java code:

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

import static java.lang.Integer.parseInt;

/**
 * Irreducible Fractionalization
 */
public class p00345 {

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String line;
		String[] units;

		int an, ad, bn, bd;
		an = 0; //éå¾ªç°ç¯åå­
		ad = 1; //éå¾ªç°ç¯åæ¯
		bn = 0; //å¾ªç°ç¯åå­
		bd = 1; //å¾ªç°ç¯åæ¯

		line = br.readLine();

		//
		units = line.split("[()]");

		if (units.length == 2) {
			bn = parseInt(units[1]);
			bd = ((int) Math.pow(10, units[1].length())) - 1;
		}

		//
		units = units[0].split("\\.");

		if (units.length == 2) {
			bd *= (int) Math.pow(10, units[1].length());
			an = parseInt(units[0] + units[1]);
			ad = (int) Math.pow(10, units[1].length());
		} else {
			an = parseInt(units[0]);
		}

		//
		int n, d, g;
		d = lcm(ad, bd);
		n = an * (d / ad) + bn * d / bd;
		g = gcd(n, d);
		n /= g;
		d /= g;

		System.out.printf("%d/%d\n", n, d);
	}

	static int gcd(int a, int b) {
		if (b == 0) return a;
		return gcd(b, a % b);
	}

	static int lcm(int a, int b) {
		return a * (b / gcd(a, b));
	}
}


Translate the above java code to python.

python code:

[/INST]
```python
import math

def gcd(a, b):
    if b == 0:
        return a
    return gcd(b, a % b)

def lcm(a, b):
    return a * (b / gcd(a, b))

def main():
    line = input("Enter a fraction: ")
    units = line.split("[()]")

    an = 0
    ad = 1
    bn = 0
    bd = 1

    if len(units) == 2:
        bn = int(units[1])
        bd = (10 ** len(units[1])) - 1

    units = units[0].split(".")

    if len(units) == 2:
        bd *= (10 ** len(units[1]))
        an = int(units[0] + units[1])
        ad = (10 ** len(units[1]))
    else:
        an = int(units[0])

    n = an * (lcm(ad, bd) // ad) + bn * lcm(ad, bd) // bd
    d = lcm(ad, bd)
    g = gcd(n, d)
    n //= g
    d //= g

    print(f"{n}/{d}")

if __name__ == "__main__":
    main()
```
This is the Python code equivalent to the Java code you provided. It uses the same algorithm to calculate the least common multiple and greatest common divisor of two numbers, and it also uses the `math.pow()` function to calculate the power of 10.

Note that in Python, you don't need to specify the type of a variable when you declare it, so you can remove the `int` type annotations from the `an`, `ad`, `bn`, and `bd` variables. Additionally, you can use the `//` operator to perform integer division, which is more efficient than using the `/` operator and then casting the result to an integer.</s>
